web_fragments/tests/test_fragment.py

```diff
@@ -0,0 +1,264 @@
+        fragment_dict = fragment_dict if fragment_dict else fragment.to_dict()
+        self.assertEqual(fragment_dict['content'], TEST_HTML)
+        self.assertEqual(fragment_dict['js_init_fn'], TEST_JS_INIT_FN)
+        self.assertEqual(fragment_dict['js_init_version'], EXPECTED_JS_INIT_VERSION)
+        self.assertEqual(fragment_dict['json_init_args'], TEST_JSON_INIT_ARGS)
+        self.assertEqual(fragment_dict['resources'], self.test_resources_dict())
+
+    def test_resources_dict(self):
+        """
+        Returns the expected test resources.
+        """
+        return [
+            {
+                'kind': 'text',
+                'data': TEST_CSS,
+                'mimetype': 'text/css',
+                'placement': 'head',
+            },
+            {
+                'kind': 'url',
+                'data': TEST_CSS_URL,
+                'mimetype': 'text/css',
+                'placement': 'head',
+            },
+            {
+                'kind': 'text',
+                'data': TEST_JS,
+                'mimetype': 'application/javascript',
+                'placement': 'foot',
+            },
+            {
+                'kind': 'url',
+                'data': TEST_JS_URL,
+                'mimetype': 'application/javascript',
+                'placement': 'foot',
+            },
+        ]
+
+    def test_to_dict(self):
+        """
+        Tests the to_dict method.
+        """
+        fragment = self.create_test_fragment()
+        fragment_dict = fragment.to_dict()
+        self.validate_fragment(fragment_dict=fragment_dict)
+
+    def test_from_dict(self):
+        """
+        Tests the from_dict method.
+        """
+        test_dict = {
+            'content': TEST_HTML,
+            'resources': self.test_resources_dict(),
+            'js_init_fn': TEST_JS_INIT_FN,
+            'js_init_version': EXPECTED_JS_INIT_VERSION,
+            'json_init_args': TEST_JSON_INIT_ARGS,
+        }
+        fragment = Fragment.from_dict(test_dict)
+        self.validate_fragment(fragment)
+
+    def test_body_html(self):
+        """
+        Tests the body_html method.
+        """
+        fragment = self.create_test_fragment()
+        html = fragment.body_html()
+        self.assertEqual(html, TEST_HTML)
+
+    def test_head_html(self):
+        """
+        Tests the head_html method.
+        """
+        fragment = self.create_test_fragment()
+        html = fragment.head_html()
+        self.assertEqual(
+            html.replace('\n', ''),
+            CSS_EXPECTED_HTML.format(css=TEST_CSS) + CSS_LINK_EXPECTED_HTML.format(css_url=TEST_CSS_URL)
+        )
+
+    def test_foot_html(self):
+        """
+        Tests the foot_html method.
+        """
+        fragment = self.create_test_fragment()
+        html = fragment.foot_html()
+        self.assertEqual(
+            html.replace('\n', ''),
+            JS_EXPECTED_HTML.format(js=TEST_JS) + JS_LINK_EXPECTED_HTML.format(js_url=TEST_JS_URL)
+        )
```

Comment:
Thanks, I agree that this is confusing. I've converted this to pytest syntax and to use in checks rather than equality. I then looked into sharing logic between the various tests but the new logic is shorter and clearer so I've left it as is.

I've left the ``CSS_EXPECTED_HTML.format(css=TEST_CSS)`` pieces because I anticipate wanting to add tests in the future which use different test CSS.